source/common/http/http2/codec_impl.cc

```diff
@@ -900,65 +905,68 @@ void ConnectionImpl::sendPendingFrames() {
-    ENVOY_CONN_LOG(debug, "setting HPACK table size to {}", connection_,
-                   http2_settings.hpack_table_size_);
-  }
-
-  if (http2_settings.max_concurrent_streams_ != NGHTTP2_INITIAL_MAX_CONCURRENT_STREAMS) {
-    iv.push_back({NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_settings.max_concurrent_streams_});
-    ENVOY_CONN_LOG(debug, "setting max concurrent streams to {}", connection_,
-                   http2_settings.max_concurrent_streams_);
-  }
-
-  if (http2_settings.initial_stream_window_size_ != NGHTTP2_INITIAL_WINDOW_SIZE) {
-    iv.push_back(
-        {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_settings.initial_stream_window_size_});
-    ENVOY_CONN_LOG(debug, "setting stream-level initial window size to {}", connection_,
-                   http2_settings.initial_stream_window_size_);
-  }
-
+void ConnectionImpl::sendSettings(
+    const envoy::config::core::v3::Http2ProtocolOptions& http2_options, bool disable_push) {
+  absl::InlinedVector<nghttp2_settings_entry, 10> settings;
+  auto insertParameter = [&settings](const nghttp2_settings_entry& entry) mutable -> bool {
+    const auto it = std::find_if(settings.cbegin(), settings.cend(),
+                                 [&entry](const nghttp2_settings_entry& existing) {
+                                   return entry.settings_id == existing.settings_id;
+                                 });
+    if (it != settings.end()) {
+      return false;
+    }
+    settings.push_back(entry);
+    return true;
+  };
+
+  // Universally disable receiving push promise frames as we don't currently support
+  // them. nghttp2 will fail the connection if the other side still sends them.
+  // TODO(mattklein123): Remove this when we correctly proxy push promise.
+  // NOTE: This is a special case with respect to custom parameter overrides in that server push is
+  // not supported and therefore not end user configurable.
   if (disable_push) {
-    // Universally disable receiving push promise frames as we don't currently support them. nghttp2
-    // will fail the connection if the other side still sends them.
-    // TODO(mattklein123): Remove this when we correctly proxy push promise.
-    iv.push_back({NGHTTP2_SETTINGS_ENABLE_PUSH, 0});
-  }
-
-  if (!iv.empty()) {
-    int rc = nghttp2_submit_settings(session_, NGHTTP2_FLAG_NONE, &iv[0], iv.size());
+    settings.push_back(
+        {static_cast<int32_t>(NGHTTP2_SETTINGS_ENABLE_PUSH), disable_push ? 0U : 1U});
+  }
+
+  for (const auto it : http2_options.custom_settings_parameters()) {
+    ASSERT(it.identifier().value() <= std::numeric_limits<uint16_t>::max());
+    const bool result =
+        insertParameter({static_cast<int32_t>(it.identifier().value()), it.value().value()});
+    if (!result) {
+      ENVOY_CONN_LOG(debug, "duplicate custom settings parameter with id {:#x}, value {}",
+                     connection_, it.identifier().value(), it.value().value());
+      continue;
```

Comment:
Sorry can this still happen? Don't you validate against this?

Reply:
I wasn't validating for inconsistencies, but I've now added it.